### DISCUSSION TITLE
MoltenVK or Honeykrisp?

### DISCUSSION THREAD

**Discussion by ovvldc** [KhronosGroup/MoltenVK#2249] [General]
Hi everyone, I noticed this post by Alyssa Rosenzweig: https://rosenzweig.io/blog/vk13-on-the-m1-in-1-month.html

She managed to get a conformant basic Vulkan 1.3 driver running on the Apple M1, calling it Honeykrisp. It doesn't support everything we want yet (transform feedback and such), probably needs a lot of performance optimisation, and probably needs more work to run on M2 and M3 and variants than a Metal-based driver. Also, can a Linux driver even be adapted for use under MacOS?

So what would be the pros and cons of using MoltenVK in an app compared to using Honeykrisp? I can imagine the points above are in play, and I haven't checked licensing (though at least one file is MIT licensed, which is very permissive). 

At the very least I hope Honeykrisp could provide inspiration to solve some of the implementation problems that MoltenVK seems to be struggling with..

**Comment by billhollings** [comment]
Thanks for posting this. Very interesting!

@alyssarosenzweig  Wowza! Very impressive work! You rock!

And kudos to those working on reverse-engineering the Mx kernels!

> Also, can a Linux driver even be adapted for use under MacOS?

The challenge is that Linux on Apple hardware is wide open, including at the hardware level, whereas macOS & iOS are proprietary and very protective. Until that changes, on those platforms, we have to go through the OS and Metal.

Having said that, I'm impressed that Honeykrisp managed to make M1 fully Vulkan-conformant. For apps that don't use the App Store, perhaps MoltenVK can get there on macOS by making more use of private API's that map to Honeykrisp functionality. Definitely worth looking into.

**Comment by DUOLabs333** [comment]
I've had the chance to look at the userspace mesa part of Honeykrisp (which is the only part MoltenVK can feasibly use --- obviously it wouldn't make much sense to try to port over the kernel driver) --- if I understand this correctly, it seems to package Vulkan commands into a streams, which it then writes to the GPU. 

I would like to work on porting at least part of this driver to macOS (to at least evaluate its potential), but there are two concerns I have: 1. what happens when something outside macOS/Metal tries to send commands to the GPU? 2. How would you even write raw commands to the GPU from macOS?

**Reply by DUOLabs333** [reply]
However, according to [this](https://social.treehouse.systems/@AsahiLinux/112564879305301735), it doesn't seem that this would be worthwhile, as any effort (w/c)ould break between upgrades, even within minor versions, so the Asahi team likely (c/w)ouldn't offer much support in this direction (I'm stuck at Big Sur for the forseeable future, so I wouldn't care much about potential instability). However, [this](https://rosenzweig.io/blog/asahi-gpu-part-4.html) implies that it is possible, but the functionality was (likely) later dropped, but I can't find the commit(s) where it was implemented.

**Comment by billhollings** [comment]
> How would you even write raw commands to the GPU from macOS?

In general, you can't. Apple intentionally sandboxes the GPU behind Metal, which it guard through the App Store and other means, and uses internal APIs that regularly change (ie. no ABI). See [this post](https://social.treehouse.systems/@AsahiLinux/112564879305301735) for more on that. So, for a standard app on macOS, it has to go through Metal.

> I would like to work on porting at least part of this driver to macOS (to at least evaluate its potential)

Coincidentally, Mesa is starting to look at exactly this, and if you're interested, you may be able to get involved there. See this [Phoronix article](https://www.phoronix.com/news/LunarG-Vulkan-Metal), and [this discussion](https://gitlab.freedesktop.org/mesa/mesa/-/issues/11990) on the Mesa website.

**Reply by DUOLabs333** [reply]
Sounds interesting, I didn't know about this. And yeah, I noted in a comment that unless you're willing to work with undocumented APIs, that directly porting over Honeykrisp seems like a non-starter.

**Comment by ovvldc** [comment]
I was hoping that Alyssa's solutions to geometry shaders and transform feedback and such could be adapted to MoltenVK. But there seems to be very little movement on those issues in general.

**Reply by DUOLabs333** [reply]
1. Large parts of Honeykrisp are not (easily) translatable to Metal, because it works on a much lower level (if the commands they use map to publicly available Metal functions, then that would be different)

2. At least for transform feedback, there was a PR, but Alyssa said it wasn't ready, an incomplete implementation would fracture the ecosystem further, etc., and the PR floundered.

**Comment by ovvldc** [comment]
I know, I followed that PR. It wasn't just that Alyssa said it wasn't ready but that the author didn't seem to have a way to address the concerns, which sounded fair to me as a non expert. I hope it gets picked up again at some point, same with geometry shaders.
